### DISCUSSION TITLE
controls_if function does not work with string values

### DISCUSSION THREAD

**Discussion by paoespinozarias** [morph-kgc/morph-kgc#142] [Q&A] [answered]
####  What Happens?
I tested this function using a string value in the `rr:template` parameter but it trhows the following exception:
![image](https://user-images.githubusercontent.com/16767406/212884349-867edb17-e98f-4e76-b2a0-335079d18d5f.png)

The output should be the same than the one obtained when running the test as it is currently configured.

#### To Reproduce
I reused the same [calendar.csv](https://github.com/morph-kgc/morph-kgc/blob/main/test/fno/controls_if/calendar.csv) file from the test but I replaced 1 by X:

```
service_id | monday | tuesday | wednesday | thursday | friday | saturday | sunday | start_date | end_date
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
4 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 6/30/2017 | 6/30/2018
5 | X | X | X | X | X | X | X | 6/30/2017 | 6/30/2018
```

I reused the same [mapping.ttl](https://github.com/morph-kgc/morph-kgc/blob/main/test/fno/controls_if/mapping.ttl) file from the test but I changed a bit the Execution statements:

```
<#Execution1>
    fnml:function grel:controls_if ;
    fnml:input
        [
            fnml:parameter grel:bool_b ;
            fnml:valueMap [
                rr:template "X == {monday}"
            ]
        ] ,
        [
            fnml:parameter grel:any_true ;
            fnml:value "true"
        ] ,
        [
            fnml:parameter grel:any_false ;
            fnml:value "false"
        ] .

<#Execution2>
    fnml:function grel:controls_if ;
    fnml:input
        [
            fnml:parameter grel:bool_b ;
            fnml:valueMap [
                rr:template "X == {tuesday}"
            ]
        ] ,
        [
            fnml:parameter grel:any_true ;
            fnml:value "true"
        ] .
```

#### Environment (please complete the following information):
I'm testing the latest version of morph-kgc installed from the GitHub repository.

**Comment by arenas-guerrero-julian** [comment]
Hi @paoespinozarias ,

Could you please try with `rr:template "'X' == {monday}"`? Note the quotes around `X`.

**Comment by paoespinozarias** [comment, marked as answer]
Hi @arenas-guerrero-julian ,

Thank you for the support. 
This configuration works but you need to also put quotes around {monday} -> `rr:template "'X' == '{monday}'"`

**Comment by arenas-guerrero-julian** [comment]
Thanks :)

This is because the expression evaluates to a Python expression. If no quotes are used, a variable (that does not exist) is assumed; using quotes it evaluates to a string.

**Comment by paoespinozarias** [comment]
Thank you for the explanation :-) 

However, I just found that my output looks different to yours [outuput.nq](https://github.com/morph-kgc/morph-kgc/blob/main/test/fno/controls_if/output.nq) file.  

My output does not contain the values annotated as the corresponding boolean datatypes: 
```
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .

<http://transport.linkeddata.es/madrid/metro/calendar_rules/4> <http://vocab.gtfs.org/terms#monday> false .

<http://transport.linkeddata.es/madrid/metro/calendar_rules/5> <http://vocab.gtfs.org/terms#monday> true ;
    <http://vocab.gtfs.org/terms#tuesday> true .
```

Should I configure something else in the mapping file?

**Comment by arenas-guerrero-julian** [comment]
Hi @paoespinozarias ,

I just tried this and it is working for me. I see that your output is not N-Triples, are you runnings as a library? what OS are you running morph-kgc with?

My maping is:

```ttl
@prefix rr: <http://www.w3.org/ns/r2rml#> .
@prefix foaf: <http://xmlns.com/foaf/0.1/> .
@prefix ex: <http://example.com/> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .
@prefix rml: <http://semweb.mmlab.be/ns/rml#> .
@prefix ql: <http://semweb.mmlab.be/ns/ql#> .
@prefix fnml: <http://semweb.mmlab.be/ns/fnml#> .
@prefix fno: <https://w3id.org/function/ontology#> .
@prefix morph-kgc: <https://github.com/morph-kgc/morph-kgc/function/built-in.ttl#> .
@prefix grel: <http://users.ugent.be/~bjdmeest/function/grel.ttl#> .
@prefix idlab-fn: <http://example.com/idlab/function/> .
@prefix gtfs: <http://vocab.gtfs.org/terms#> .

@base <http://example.com/base/> .

<TriplesMap1>
    rml:logicalSource [
        rml:source "calendar.csv" ;
        rml:referenceFormulation ql:CSV ;
    ];
    rr:subjectMap [
        rr:template "http://transport.linkeddata.es/madrid/metro/calendar_rules/{service_id}"
    ];
    rr:predicateObjectMap [
        rr:predicate gtfs:monday;
        rr:objectMap [
            fnml:execution <#Execution1> ;
            rr:datatype xsd:boolean
        ]
    ];
    rr:predicateObjectMap [
        rr:predicate gtfs:tuesday;
        rr:objectMap [
            fnml:execution <#Execution2> ;
            rr:datatype xsd:boolean
        ]
    ].

<#Execution1>
    fnml:function grel:controls_if ;
    fnml:input
        [
            fnml:parameter grel:bool_b ;
            fnml:valueMap [
                rr:template "'X' == '{tuesday}'"
            ]
        ] ,
        [
            fnml:parameter grel:any_true ;
            fnml:value "true"
        ] ,
        [
            fnml:parameter grel:any_false ;
            fnml:value "false"
        ] .

<#Execution2>
    fnml:function grel:controls_if ;
    fnml:input
        [
            fnml:parameter grel:bool_b ;
            fnml:valueMap [
                rr:template "'X' == '{tuesday}'"
            ]
        ] ,
        [
            fnml:parameter grel:any_true ;
            fnml:value "true"
        ] .

```

**Comment by paoespinozarias** [comment]
Hi @arenas-guerrero-julian!

I am using it as a library, running on Windows 10 Enterprise.  

I just serialized the output in N-Triple format and it works for me. I got the same output than yours using `g_morph.serialize("output.nq", format='nt').`  However, shouldn't the output be well annotated for TTL serialization too (`g_morph.serialize("output.nq", format='turttle')`)? We have been serializing in TTL all our KGs built by morph-kgc and everything was well annotated so far. 

Could you please try serializing your output in TTL and see if it works for you?

**Reply by arenas-guerrero-julian** [reply]
Thanks @paoespinozarias 

As you are executing as a library, the serialization depends on RDFlib, not morph-kgc. However, looking at the [Turtle spec](https://www.w3.org/TR/turtle/#h4_booleans) this seems like a valid syntax for booleans in turtle format (example from the spec):

```ttl
@prefix : <http://example.org/stats> .
<http://somecountry.example/census2007>
    :isLandlocked false .           # xsd:boolean
```

If, for some reason you need n3 syntax, maybe you can use the `materialize_oxigraph` method of morph-kgc and serialize with oxigraph. Possibly it uses n3 syntax for booleans.

Julián

**Reply by paoespinozarias** [reply]
Oh I see, thank you so much for your help.
